File: cses/forbidden_cities_block_cut.py
"""
CSES 1705 - Forbidden Cities: block-cut tree + binary-lifting LCA
https://cses.fi/problemset/task/1705

PROBLEM AND IDEA
Each query asks whether a and b remain connected after city c is deleted.
Running a fresh graph search per query would take O(q * (n + m)). Instead,
turn the graph's articulation structure into a tree. The query then becomes:
does city c lie on the unique tree path between cities a and b?

TECHNIQUES
Recursive Tarjan DFS, low-link values, an edge stack, vertex-biconnected
components, an expanded block-cut tree, recursive tree DFS, subtree intervals,
and lowest common ancestors (LCA) with binary lifting.

1. EXTRACT BICONNECTED BLOCKS
   discovered[v] is the discovery time of v. low[v] is the earliest discovery
   time reachable from v's DFS subtree via a back edge, also allowing v itself.
   During recursive DFS, push tree edges and back edges to strict ancestors
   onto pending_edges. When a child u returns:
       low[v] = min(low[v], low[u]).
   If low[u] >= discovered[v], this branch cannot reach a strict ancestor of
   v. Pop edges through the tree edge (v, u), inclusive: these edges form one
   biconnected block. Equality is included because reaching v itself does not
   bypass v if it is forbidden. Bridges become blocks with two vertices.
   Each edge is pushed and popped at most once. Edge IDs distinguish parallel
   roads; self-loops have no effect on connectivity and need no block.

2. KEEP EVERY ORIGINAL CITY IN THE BLOCK-CUT TREE
   Original cities keep indices 0..n-1. For each block, append one new node
   connected to every distinct city in that block. A last_block array prevents
   duplicate links when several extracted edges share an endpoint.

   This is an expanded form of the usual block-cut tree: articulation cities
   join multiple blocks, while non-articulation cities remain as leaves of
   their sole block. It avoids both a separate city-to-block ID mapping and
   explicit articulation-point flags.

   Example: triangles {1,2,3} and {3,4,5} produce two block nodes connected
   through city 3. Cities 1 and 2 attach to the first block, and 4 and 5 to
   the second. Removing city 3 blocks 1 -> 5 but still allows 1 -> 2.

   Because the original graph is connected, this incidence structure is a
   tree. If B blocks are extracted, there are T = n + B nodes and T - 1 tree
   edges. Each block closes at a distinct DFS-tree edge, so B <= n - 1 and
   T <= 2n - 1. For one city, the tree consists of that city alone.

3. PREPROCESS ANCESTORS WITH A SECOND RECURSIVE DFS
   Root the tree at city 0. Compute depth, parent, and subtree discovery-time
   intervals [entered[v], exited[v]). Vertex v is an ancestor of x exactly
   when entered[v] <= entered[x] < exited[v].
   Build up[j][v] = the ancestor of v after 2**j parent steps. Set the root's
   parent to itself. Use T.bit_length() levels, based on the constructed tree,
   and access only levels 0 through len(up)-1. Binary lifting finds an LCA in
   O(log T). Compact array('i') rows reduce the ancestor table's memory use.

4. USE ONE LCA TO TEST THE FORBIDDEN CITY
   Let w = LCA(a, b). The path a -> b consists of the chains a -> w and b -> w.
   Thus c lies on the path precisely when BOTH conditions hold:
   - c is an ancestor of a or an ancestor of b;
   - depth[c] >= depth[w], so c is not above the path's highest vertex.
   If the first condition fails, skip the LCA calculation altogether.
   Answer NO if c is on the path, and YES otherwise. Forbidden endpoints are
   rejected immediately; a == b != c permits the zero-edge route.

CORRECTNESS
Within a biconnected block, removing one vertex leaves all surviving vertices
connected. A bridge block also satisfies this statement: deleting an endpoint
leaves just one vertex. Blocks can meet only at articulation points, so graph
connectivity after deleting c matches connectivity among the city nodes after
deleting c from the expanded block-cut tree. A tree has a unique path between
two vertices; removing c destroys that connection exactly when c is on the
path. The ancestor-and-LCA condition checks exactly those path vertices.

COMPLEXITY AND RECURSIVE PYTHON
Block construction: O(n + m) time and space.
LCA preprocessing: O(T log T) time and space.
Queries: O(log T) each, at most O(log n) for n > 1.
Total: O(m + n log n + q log n) time and O(m + n log n + q) space, including
the buffered answers. The n = 1 case takes constant preprocessing/query time.

Both graph traversals use actual recursive calls. A chain of n cities becomes
a tree of 2n - 1 nodes, so sys.setrecursionlimit(2*n + 1000) provides enough
Python call depth. On interpreters where Python recursion uses native stack
space, that limit alone is insufficient. The solver therefore runs in one
worker thread with a requested 256 MiB stack. This thread provides stack
capacity; it does not parallelize the algorithm. The original graph is freed
before the ancestor table is allocated. pending_edges is the component edge
stack required by Tarjan's algorithm, not an iterative DFS implementation.
"""

import sys
import threading
from array import array


class BlockCutBuilder:
    def __init__(self, n):
        self.graph = [[] for _ in range(n)]
        self.roads = []
        self.tree = [[] for _ in range(n)]
        self.discovered = [-1] * n
        self.low = [0] * n
        self.pending_edges = []
        self.last_block = [-1] * n
        self.timer = 0

    def add_road(self, a, b):
        edge = len(self.roads)
        self.roads.append((a, b))
        self.graph[a].append((b, edge))
        self.graph[b].append((a, edge))

    def _finish_block(self, first_edge):
        block = len(self.tree)
        self.tree.append([])
        while True:
            edge = self.pending_edges.pop()
            for city in self.roads[edge]:
                if self.last_block[city] != block:
                    self.last_block[city] = block
                    self.tree[city].append(block)
                    self.tree[block].append(city)
            if edge == first_edge:
                break

    def _visit(self, v, incoming_edge):
        self.discovered[v] = self.low[v] = self.timer
        self.timer += 1
        for u, edge in self.graph[v]:
            if edge == incoming_edge:
                continue
            if self.discovered[u] == -1:
                self.pending_edges.append(edge)
                self._visit(u, edge)
                if self.low[u] < self.low[v]:
                    self.low[v] = self.low[u]
                if self.low[u] >= self.discovered[v]:
                    self._finish_block(edge)
            elif self.discovered[u] < self.discovered[v]:
                # The reverse direction would duplicate this back edge.
                self.pending_edges.append(edge)
                if self.discovered[u] < self.low[v]:
                    self.low[v] = self.discovered[u]

    def build(self):
        self._visit(0, -1)
        return self.tree


class TreePaths:
    def __init__(self, tree):
        self.tree = tree
        size = len(tree)
        self.entered = [0] * size
        self.exited = [0] * size
        self.depth = [0] * size
        self.up = [array("i", [0]) * size]
        self.timer = 0
        self._visit(0, 0)
        for _ in range(1, size.bit_length()):
            previous = self.up[-1]
            self.up.append(array("i", (previous[previous[v]] for v in range(size))))

    def _visit(self, v, parent):
        self.entered[v] = self.timer
        self.timer += 1
        self.up[0][v] = parent
        for u in self.tree[v]:
            if u != parent:
                self.depth[u] = self.depth[v] + 1
                self._visit(u, v)
        self.exited[v] = self.timer

    def _lca(self, a, b):
        entered = self.entered
        exited = self.exited
        if entered[a] <= entered[b] < exited[a]:
            return a
        if entered[b] <= entered[a] < exited[b]:
            return b
        for ancestors in reversed(self.up):
            candidate = ancestors[a]
            if not (entered[candidate] <= entered[b] < exited[candidate]):
                a = candidate
        return self.up[0][a]

    def on_path(self, a, b, c):
        entered = self.entered
        exited = self.exited
        if not (entered[c] <= entered[a] < exited[c]
                or entered[c] <= entered[b] < exited[c]):
            return False
        return self.depth[c] >= self.depth[self._lca(a, b)]


def main():
    read = sys.stdin.buffer.readline
    first_line = read()
    if not first_line:
        return
    n, m, q = map(int, first_line.split())
    sys.setrecursionlimit(2 * n + 1000)

    builder = BlockCutBuilder(n)
    for _ in range(m):
        a, b = map(int, read().split())
        builder.add_road(a - 1, b - 1)
    tree = builder.build()
    del builder  # The original adjacency lists and component buffers are done.
    paths = TreePaths(tree)

    answers = []
    for _ in range(q):
        a, b, c = map(int, read().split())
        a -= 1
        b -= 1
        c -= 1
        blocked = a == c or b == c or paths.on_path(a, b, c)
        answers.append("NO" if blocked else "YES")
    sys.stdout.write("\n".join(answers) + "\n")


if __name__ == "__main__":
    threading.stack_size(256 * 1024 * 1024)
    worker = threading.Thread(target=main)
    worker.start()
    worker.join()

File: cses/forbidden_cities_block_cut.cpp
/*
CSES 1705 - Forbidden Cities: block-cut tree + binary-lifting LCA
https://cses.fi/problemset/task/1705

PROBLEM AND STRATEGY
For each query (a, b, c), decide whether a and b remain connected after city c
is removed. Searching the graph for every query is too expensive. Instead,
describe the graph's articulation structure as a tree and test whether c lies
on the unique tree path between a and b.

TECHNIQUES
Recursive Tarjan DFS, discovery times and low-link values, an edge stack,
vertex-biconnected components, a block-cut tree, DFS subtree intervals, and
lowest common ancestors (LCA) with binary lifting.

1. FIND BICONNECTED BLOCKS WITH AN EDGE STACK
   discovered[v] records when DFS first visits v. low[v] is the smallest
   discovery time reachable from v's DFS subtree using a back edge, also
   allowing v itself. Push every tree edge and every back edge to an ancestor
   onto pending_edges. Each undirected edge is pushed at most once.

   After visiting a DFS child u of v, low[u] >= discovered[v] means that u's
   branch cannot reach a strict ancestor of v. This closes a biconnected block:
   pop edges through the tree edge (v, u), inclusive. Their endpoints are the
   vertices of that block. The >= matters because reaching v itself does not
   bypass v when it is removed. A bridge forms a block with two vertices.
   Edge IDs allow parallel roads; self-loops can be ignored for connectivity.

2. BUILD AN EXPANDED BLOCK-CUT TREE
   Keep all original cities as nodes 0..n-1. For every block, append a new node
   and connect it to every distinct city in that block. last_block[city]
   prevents adding a city twice when several popped edges share an endpoint.

   In the usual compressed block-cut tree, only articulation points keep
   their own nodes; other cities map to their block. Here those other cities
   remain as leaves. This equivalent representation needs no city-ID mapping
   and no separate articulation-point array.

   For two triangles {1,2,3} and {3,4,5}, city 3 links the two block nodes;
   cities 1,2 are leaves of the first block and 4,5 of the second. Removing 3
   separates 1 from 5, but does not separate 1 from 2.

   The input graph is connected, so the incidence structure is a tree. If
   there are B blocks, it has T = n + B nodes and T - 1 edges. Since every
   block closes at a distinct DFS-tree edge, B <= n - 1. A single city simply
   produces a one-node tree.

3. PREPROCESS TREE PATHS WITH A SECOND RECURSIVE DFS
   Root the new tree at city 0. Record depth, direct parent, and the half-open
   subtree interval [entered[v], exited[v]). Then:
       ancestor(v, x) iff entered[v] <= entered[x] < exited[v].
   Build up[j][v], the ancestor of v after 2^j steps. The number of levels is
   computed from T, not from the original graph size. Valid indices are
   0..levels-1. Binary lifting finds LCA(a, b) in O(log T).

4. ANSWER WITH ONE LCA QUERY
   Let w = LCA(a, b). The a-b path is the union of the ancestor chains from
   a to w and b to w. Therefore c lies on that path exactly when:
       (ancestor(c, a) or ancestor(c, b)) and depth[c] >= depth[w].
   If c is not an ancestor of either endpoint, no LCA call is needed.
   Print NO when c is on the path, YES otherwise. Rejecting a == c or b == c
   explicitly also handles forbidden endpoints. If a == b != c, answer YES.

WHY THE REDUCTION IS CORRECT
A biconnected block stays connected between its surviving vertices after
one city is removed; for a bridge block, at most one vertex survives if an
endpoint is removed. Distinct blocks meet only at articulation points. Thus
a graph route avoiding c exists exactly when its endpoint cities remain
connected in the expanded block-cut tree with node c removed. In a tree,
removing a node separates two other nodes precisely when it lies on their
unique path. The LCA condition checks exactly this property.

COMPLEXITY AND RECURSION
Block construction takes O(n + m) time and space. LCA preprocessing takes
O(T log T) time and space, and each query takes O(log T) in the worst case.
With T <= 2n - 1, total time is O(m + n log n + q log n), and space is
O(m + n log n). For n = 1, preprocessing and each query are constant time.
Both DFS traversals are recursive. A chain can produce nearly 2n recursive
tree calls, so on POSIX systems the program requests a larger process stack,
up to the system's hard limit. pending_edges is Tarjan's component stack;
it does not replace recursive DFS calls.
*/

#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

#if defined(__unix__) || defined(__APPLE__)
#include <sys/resource.h>
#endif

using namespace std;

class BlockCutBuilder {
    vector<vector<pair<int, int>>> graph;
    vector<pair<int, int>> roads;
    vector<vector<int>> tree;
    vector<int> discovered, low, pending_edges, last_block;
    int timer = 0;

    void finish_block(int first_edge) {
        int block = static_cast<int>(tree.size());
        tree.emplace_back();
        while (true) {
            int edge = pending_edges.back();
            pending_edges.pop_back();
            auto [a, b] = roads[edge];
            for (int city : {a, b}) {
                if (last_block[city] != block) {
                    last_block[city] = block;
                    tree[city].push_back(block);
                    tree[block].push_back(city);
                }
            }
            if (edge == first_edge) break;
        }
    }

    void visit(int v, int incoming_edge) {
        discovered[v] = low[v] = timer++;
        for (auto [u, edge] : graph[v]) {
            if (edge == incoming_edge) continue;
            if (discovered[u] == -1) {
                pending_edges.push_back(edge);
                visit(u, edge);
                low[v] = min(low[v], low[u]);
                if (low[u] >= discovered[v]) finish_block(edge);
            } else if (discovered[u] < discovered[v]) {
                // Process a back edge only from its descendant endpoint.
                pending_edges.push_back(edge);
                low[v] = min(low[v], discovered[u]);
            }
        }
    }

public:
    explicit BlockCutBuilder(int n)
        : graph(n), tree(n), discovered(n, -1), low(n), last_block(n, -1) {}

    void add_road(int a, int b) {
        int edge = static_cast<int>(roads.size());
        roads.emplace_back(a, b);
        graph[a].emplace_back(b, edge);
        graph[b].emplace_back(a, edge);
    }

    vector<vector<int>> build() {
        visit(0, -1);
        return std::move(tree);
    }
};

class TreePaths {
    const vector<vector<int>>& tree;
    vector<vector<int>> up;
    vector<int> entered, exited, depth;
    int timer = 0;

    void visit(int v, int parent) {
        entered[v] = timer++;
        up[0][v] = parent;
        for (int u : tree[v]) {
            if (u == parent) continue;
            depth[u] = depth[v] + 1;
            visit(u, v);
        }
        exited[v] = timer;
    }

    bool ancestor(int v, int x) const {
        return entered[v] <= entered[x] && entered[x] < exited[v];
    }

    int lca(int a, int b) const {
        if (ancestor(a, b)) return a;
        if (ancestor(b, a)) return b;
        for (int level = static_cast<int>(up.size()) - 1; level >= 0; --level) {
            int candidate = up[level][a];
            if (!ancestor(candidate, b)) a = candidate;
        }
        return up[0][a];
    }

public:
    explicit TreePaths(const vector<vector<int>>& input_tree)
        : tree(input_tree), entered(tree.size()), exited(tree.size()),
          depth(tree.size()) {
        int size = static_cast<int>(tree.size());
        int levels = 1;
        while ((1LL << levels) <= size) ++levels;
        up.assign(levels, vector<int>(size));
        visit(0, 0);
        for (int level = 1; level < levels; ++level) {
            for (int v = 0; v < size; ++v) {
                up[level][v] = up[level - 1][up[level - 1][v]];
            }
        }
    }

    bool on_path(int a, int b, int c) const {
        if (!ancestor(c, a) && !ancestor(c, b)) return false;
        return depth[c] >= depth[lca(a, b)];
    }
};

void allow_deep_recursion() {
#if defined(__unix__) || defined(__APPLE__)
    rlimit limit;
    if (getrlimit(RLIMIT_STACK, &limit) == 0) {
        rlim_t wanted = 128ULL * 1024 * 1024;
        if (limit.rlim_max != RLIM_INFINITY) wanted = min(wanted, limit.rlim_max);
        if (limit.rlim_cur != RLIM_INFINITY && limit.rlim_cur < wanted) {
            limit.rlim_cur = wanted;
            setrlimit(RLIMIT_STACK, &limit);
        }
    }
#endif
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    allow_deep_recursion();

    int n, m, q;
    cin >> n >> m >> q;

    vector<vector<int>> tree;
    {
        BlockCutBuilder builder(n);
        for (int i = 0; i < m; ++i) {
            int a, b;
            cin >> a >> b;
            builder.add_road(a - 1, b - 1);
        }
        tree = builder.build();
    } // Release the original graph before allocating the ancestor table.
    TreePaths paths(tree);

    while (q--) {
        int a, b, c;
        cin >> a >> b >> c;
        --a;
        --b;
        --c;
        bool blocked = a == c || b == c || paths.on_path(a, b, c);
        cout << (blocked ? "NO\n" : "YES\n");
    }
    return 0;
}
